File: src/ros_tools/scripts/ground_camera_node.py
#!/usr/bin/env python
import rospy
from sensor_msgs.msg import Image
from cv_bridge import CvBridge
import cv2


def ground_camera_node():
    # Initialize the ROS node
    rospy.init_node("ground_camera_node", anonymous=True)

    # Create a publisher for the camera topic
    camera_pub = rospy.Publisher("/camera/ground", Image, queue_size=1)

    # Create a CvBridge object
    bridge = CvBridge()

    # Open the ground camera device
    ground_camera = cv2.VideoCapture("/dev/ground")

    # Set the camera frame width and height
    ground_camera.set(cv2.CAP_PROP_FRAME_WIDTH, 640)
    ground_camera.set(cv2.CAP_PROP_FRAME_HEIGHT, 480)

    # Main loop
    rate = rospy.Rate(30)  # 30 Hz
    while not rospy.is_shutdown():
        # Read a frame from the ground camera
        ret, frame = ground_camera.read()

        if ret:
            # Convert the frame to a ROS Image message
            image_msg = bridge.cv2_to_imgmsg(frame, encoding="bgr8")

            # Publish the image message
            camera_pub.publish(image_msg)

        rate.sleep()

    # Release the ground camera
    ground_camera.release()


if __name__ == "__main__":
    try:
        ground_camera_node()
    except rospy.ROSInterruptException:
        pass

File: src/ros_tools/src/ground_camera_node.cpp
#include <cv_bridge/cv_bridge.h>
#include <opencv2/opencv.hpp>
#include <ros/ros.h>
#include <sensor_msgs/Image.h>

int main(int argc, char **argv) {
    // 初始化ROS节点
    ros::init(argc, argv, "ground_camera_node");
    ros::NodeHandle nh;

    // 创建发布器
    ros::Publisher camera_pub =
        nh.advertise<sensor_msgs::Image>("/camera/ground", 10);

    // 打开摄像头设备
    cv::VideoCapture ground_camera("/dev/ground", cv::CAP_V4L2);
    if (!ground_camera.isOpened()) {
        ROS_ERROR("无法打开摄像头");
        return -1;
    }

    // 设置摄像头分辨率
    ground_camera.set(cv::CAP_PROP_FRAME_WIDTH, 640);
    ground_camera.set(cv::CAP_PROP_FRAME_HEIGHT, 480);
    ground_camera.set(cv::CAP_PROP_FPS, 30);

    cv::Mat frame;
    cv_bridge::CvImage cv_image;
    cv_image.encoding = "bgr8";

    // 主循环
    while (ros::ok()) {
        ground_camera >> frame;
        // 转换帧为ROS图像消息
        cv_image.image = frame;
        sensor_msgs::Image ros_image;
        cv_image.toImageMsg(ros_image);
        // 发布图像消息
        camera_pub.publish(ros_image);
    }

    // 释放摄像头
    ground_camera.release();

    return 0;
}